### PR TITLE
Refurbish Rose Docs

### LINKED ISSUE #2503: Update rose-macro tutorial
On reading the rose macro tutorial whilst reviewing #2494  the following issues came to light:

- [x] Subprocess in example macro doesn't work because it returns a byte type at Python 3
- [x] Web getter no longer works
- [x] Node string not evaluated :. "'Earth'" != 'Earth'
- [x] References to Rose-edit.

### LINKED ISSUE #2504: Python 3/Rose 2 General Checks on Documentation
Update the documentation - including the following items:

- [x] Look for refs to ``rose`` not ``metomi.rose`` in code examples.
- [x] Look for Python2 Shebang Lines in examples.

### PR DESCRIPTION
**Pull request by wxtim**
Closes #2504 
Closes #2503 

Read through entire user guide section, except "further topics".

- [x] Removed cheat sheet because it is entirely about things which aren't in the Cylc 8 Ecosystem.
- [x] Removed references to `rose suite-run`, replaced with notes on Cylc scripts where possible.
- [x] Removed explanation of Suite Dir vs. Rundir as this is now `cylc install`'s problem.
- [x] Re-written tutorial about Rose Suites. Mostly passed over much more quickly with links to in-depth docs for Cylc Validate/Install/Play, leaving material about Rose configs.
- [x] Removed references to `rose host-select`. It still exists, but we want people to use platform selection.
- [x] Referred to `[template variables]` section rather than `[jinja2:suite.rc]` wherever possible.
- [x] Checked the further topics tutorials.
- [x] Fixed a [small Python 3](https://stackoverflow.com/questions/11941817/how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error)  bug in validation script.
- [x] Grepped for dead-ends from `metomi/rose/rose.py` in the docs*

\* See
```
cd ~/rose/sphinx
for deadend in $(python -c "from metomi.rose.rose import DEAD_ENDS; x = ['.'.join(end) for end in DEAD_ENDS]; [print(i) for i in x]"); do 
  echo $deadend
  grep -rin "$deadend" *
done
``` 

--------------

### Additional changes after review 1 from @oliver-sanders 
- [x] Searched for instances of `cylc gui` and changed if req'd.
- [x]  `cylc get-config` → `cylc config`
- [x] `[cylc]` → `[scheduler]`
- [x] Ran a crude check for the validity of Cylc code snippets.

### CODE REVIEW

File: sphinx/tutorial/rose/suites.rst [hunk 1]
```diff
@@ -60,221 +86,49 @@ A Rose suite configuration is a Cylc :term:`run directory` containing a
...
-       [[dependencies]]
-           graph = hello_{{WORLD}}
+       [[graph]]
+           R1 = hello_{{WORLD}}
 
    [runtime]
        [[hello_{{WORLD}}]]
            script = echo "$GREETING {{WORLD}}"
 
-
```
Comment (wxtim) [inline sphinx/tutorial/rose/suites.rst:77]: This is all dead post the introduction of Cylc install.

File: sphinx/tutorial/rose/suites.rst [hunk 2]
```diff
@@ -4,8 +4,25 @@
...
+
+   To check:
+
+   .. code-block:: bash
+
+      $ cylc version --long
+      8.0.0 (/path/to/install)
+
+      Plugins:
+            cylc-rose       1.0.0   /net/home/h02/tpilling/metomi/cylc-rose
```
Comment (datamel) [inline sphinx/tutorial/rose/suites.rst:24]: Should change this to something anonymous.
```suggestion
            cylc-rose       1.0.0   ~/cylc-rose
```

File: sphinx/tutorial/rose/suites.rst [hunk 3]
```diff
@@ -18,7 +35,7 @@ be used to do the same for a :term:`workflow`.
 Configuration Format
 --------------------
 
-A Rose suite configuration is a Cylc :term:`run directory` containing a
+A Rose suite configuration is a Cylc :term:`source directory` containing a
```
Comment (datamel) [inline sphinx/tutorial/rose/suites.rst:38]: Could be a run directory too, no?

**Review by datamel — COMMENTED** [review]
I have read and built the docs. Happy with the change although think it best to anonymize the directory for security.

File: sphinx/tutorial/rose/suites.rst (hunk 2 above)
Comment (wxtim) [reply sphinx/tutorial/rose/suites.rst:24]: oops.
Bit late now, it's in the history.

File: sphinx/tutorial/rose/suites.rst (hunk 3 above)
Comment (wxtim) [reply sphinx/tutorial/rose/suites.rst:38]: Yes. But...
I don't think that we want to encourage that way of working.

**Comment by wxtim** [comment]
Added the reference section. Hardly any changes in https://github.com/metomi/rose/pull/2519/commits/9f5e86db322c17d0848df112d6e79caf258669ae

**Comment by wxtim** [comment]
@datamel - I've added a fair bit more, but the scope of the remaining required changes seemed smallish, so I haven't created a new PR. Sorry.

File: sphinx/api/command-reference.rst [hunk 4]
```diff
@@ -28,7 +28,7 @@ rose config-edit
 rose suite-run
 ^^^^^^^^^^^^^^
 
-This command has been replaced by ``cylc install ; cylc play``.
+This command has been replaced by ``cylc install; cylc validate ; cylc play``.
```
Comment (oliver-sanders) [inline sphinx/api/command-reference.rst:31]:
```suggestion
This command has been replaced by ``cylc validate; cylc install; cylc play``.
```

File: sphinx/tutorial/rose/furthertopics/macro-development.rst [hunk 5]
```diff
@@ -90,16 +90,20 @@ Open ``planet.py`` in a text editor and paste in the following code:
 
 .. code-block:: python
 
-   #!/usr/bin/env python2
-   # -*- coding: utf-8 -*-
+   #!/usr/bin/env python3
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/furthertopics/macro-development.rst:93]: Could do with a note explaining that this is python2/3 compatible code in order to be rose1/2 compatible in order to use the rose 1 Gui.

I'm not sure whether this shebang is needed? Does the tutorial execute this file `chmod +x <file>; ./file`, if not can remove.

File: sphinx/tutorial/rose/furthertopics/macro-development.rst [hunk 6]
```diff
@@ -90,16 +90,20 @@ Open ``planet.py`` in a text editor and paste in the following code:
...
    import re
    import subprocess
 
-   import metomi.rose.macro
+   try:
+       from metomi.rose.macro import MacroBase
+       ROSEV = 2
+   except ImportError:
+       from rose.macro import MacroBase
+       ROSEV = 1
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/furthertopics/macro-development.rst:103]: ROSEV might be a little cryptic, `PY_3 = True` might be a little clearer.

File: sphinx/tutorial/rose/furthertopics/macro-development.rst [hunk 7]
```diff
@@ -171,7 +177,7 @@ at the top of the class, like this:
 
 .. code-block:: python
 
-   class PlanetChecker(rose.macro.MacroBase):
+   class PlanetChecker(metomi.rose.macro.MacroBase):
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/furthertopics/macro-development.rst:180]: Not Rose 1 compatible.

File: sphinx/tutorial/rose/furthertopics/macro-development.rst [hunk 8]
```diff
@@ -259,39 +271,9 @@ Results
...
-   :align: center
-   :alt: Screenshot of setting with an error detected by a Rose macro.
-
-Try changing the value of ``env=WORLD`` to other solar system planets
-and re-running the macro.
-
-You can also run your macro from the command line::
+Run your macro from the command line::
 
    rose macro planet.PlanetChecker
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/furthertopics/macro-development.rst:278]:
```suggestion
   rose macro planet.PlanetChecker

Try changing the value of ``env=WORLD`` to other solar system planets
and re-running the macro.
```

File: sphinx/tutorial/rose/furthertopics/rose-bunch.rst [hunk 9]
```diff
@@ -135,8 +135,12 @@ directory of the lander app and running::
...
-:rose:file:`rose-suite.conf` files can be found) and run
-:ref:`command-rose-suite-run`.
+:rose:file:`rose-suite.conf` files can be found) and validate, install and run
+your suite::
+
+   cylc validate rose-arch
+   cylc install rose-arch
+   cylc play rose-arch
 
 Your suite should run, launch the Cylc GUI and successfully run the ``lander``
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/furthertopics/rose-bunch.rst:145]: Might need to search for "Cylc GUI" too.

File: sphinx/tutorial/rose/suites.rst [hunk 10]
```diff
@@ -4,8 +4,25 @@
...
 
-Rose Suite Configurations
-=========================
+Rose Configurations for Workflows
+=================================
+
+.. note::
+
+   The following documentation reflects installing and running a Cylc
+   workflow, and assumes that you have Cylc and the
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:13]: `s/Cylc/Cylc Flow/`

Would be good to link to cylc-flow.

File: sphinx/tutorial/rose/suites.rst [hunk 11]
```diff
@@ -4,8 +4,25 @@
...
+   cylc-rose plugin installed.
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:14]: Would be good to link to cylc-rose.

File: sphinx/tutorial/rose/suites.rst [hunk 12]
```diff
@@ -4,8 +4,25 @@
 
 .. _tutorial-rose-suites:
 
-Rose Suite Configurations
-=========================
+Rose Configurations for Workflows
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:7]: We weren't planning on renaming these.

File: sphinx/tutorial/rose/suites.rst [hunk 13]
```diff
@@ -32,6 +49,8 @@ A Rose suite configuration is a Cylc :term:`run directory` containing a
 
    :rose:conf:`rose-suite.conf[env]`
       Environment variables for use by the whole suite.
+   :rose:conf:`rose-suite.conf[template variables]`
+      Generic variables for use in the ``flow.cylc`` file.
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:53]: I would be tempted to remove the separate Jinja2 / EmPy sections and revise the note to explain that there used to be separate sections for them. (want new people to use the best practice):

```rst
   :rose:conf:`rose-suite.conf[template variables]`
      Generic variables for use by Jinja2 / EmPy in the ``flow.cylc`` file.
```

File: sphinx/tutorial/rose/suites.rst [hunk 14]
```diff
@@ -60,221 +86,49 @@ A Rose suite configuration is a Cylc :term:`run directory` containing a
...
-
-   :ref:`command-rose-suite-run` installs suites to the run directory
-   incrementally so if you change a file and restart the suite using
-   ``rose suite-run --restart`` only the changed file will be re-installed.
-   This process is strictly constructive i.e. any files deleted in the suite
-   directory will *not* be removed from the run directory. To force
-   :ref:`command-rose-suite-run` to perform a complete rebuild, use the
-   ``--new`` option.
+   # Assuming that the example above was developed in ~/cylc-src/my-workflow
+   cylc validate my-workflow    # Checks that the workflow + configuration are valid
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:128]: Easy to mistake workflow config with App config.

```suggestion
   cylc validate my-workflow    # Checks that the workflow configuration is valid
```

File: sphinx/tutorial/rose/suites.rst [hunk 15]
```diff
@@ -330,14 +183,25 @@ See the :ref:`Cheat Sheet` for more information.
...
+   #. **Tell the workflow what language to use when templating**
+
+      Add a hashbang line to the flow.cylc file to tell it to use Jinja2 to
+      process template variables:
+
+      .. code-block:: diff
+
+         + #!jinja2
+         [cylc]
+             UTC mode = True
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:204]:
```suggestion
         + #!jinja2
           [cylc]
               UTC mode = True
```

File: sphinx/tutorial/rose/suites.rst [hunk 16]
```diff
@@ -414,11 +278,14 @@ See the :ref:`Cheat Sheet` for more information.
...
+
+         cylc validate rose-suite-tutorial
 
-         rose suite-run --local-install-only
+      You can then install the workflow with :ref:`cylc install <Install-Workflow>` -
+
+         cylc install rose-suite-tutorial
 
       Inspect the installed suite, which you will find in
       the :term:`run directory`, i.e::
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:291]: The i.e. doesn't cover the numbered runs. Perhaps add runN?

File: sphinx/tutorial/rose/suites.rst [hunk 17]
```diff
@@ -663,9 +534,11 @@ Rose Applications In Rose Suite Configurations
 
       Install, validate and run the suite::
 
-         rose suite-run
+         cylc install rose-suite-tutorial
+         cylc validate rose-suite-tutorial
+         cylc play rose-suite-tutorial
+         cylc gui rose-suite-tutorial
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/suites.rst:540]: The `cylc gui` command doesn't take an argument ATM.

```suggestion
```

File: sphinx/tutorial/rose/summary.rst [hunk 18]
```diff
@@ -64,7 +64,7 @@ So far we have covered:
 
 .. ifnotslides::
 
-   Cylc suites can have Rose applications. These are stored in an ``app``
+   Cylc workflows can have Rose applications. These are stored in an ``app``
    directory and are configured using a :rose:file:`rose-app.conf` file.
 
 .. ifslides::
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/summary.rst:70]: Just spotted `cylc get-config` (now `cylc config`) below:

File: sphinx/tutorial/rose/summary.rst [hunk 19]
```diff
@@ -125,34 +125,11 @@ Suite Commands
       Runs a Rose application.
    :ref:`command-rose-task-run`
```
Comment (oliver-sanders) [inline sphinx/tutorial/rose/summary.rst:126]: Looks like a pre-existing syntax error :(

```suggestion
  :ref:`command-rose-task-run`
```

**Review by oliver-sanders — COMMENTED** [review]
Good riddance to a lot of cruft!

File: sphinx/tutorial/rose/furthertopics/macro-development.rst (hunk 6 above)
Comment (wxtim) [reply sphinx/tutorial/rose/furthertopics/macro-development.rst:103]: Ay. Done

**Comment by oliver-sanders** [comment]
(please squash when you get the chance)

**Comment by wxtim** [comment]
> (please squash when you get the chance)

History re-written like a Pharoh (We keep winning battles, just closer and closer to home).

**Comment by wxtim** [comment]
@oliver-sanders - I think that I've now responded to you. The only thing where I disagree is:

> The i.e. doesn't cover the numbered runs. Perhaps add runN?

Numbered runs are a purely Cylc implementation detail. All the Rose stuff works without needing to mention them explicitly. _If_ you do this tutorial without reading _anything_ about Cylc 8, and _if_ you then go poking the run-dir in detail this will be confusing. 

Happy to include a note, but I think adding stuff about this to the main text is distracting.

File: metomi/rose/upgrade.py [hunk 20]
```diff
@@ -153,7 +153,7 @@ def _get_config_resources(self):
         file_map = {}
         file_map[MACRO_UPGRADE_RESOURCE_FILE_ADD] = add_path
         file_map[MACRO_UPGRADE_RESOURCE_FILE_REMOVE] = rem_path
-        for key, path in file_map.items():
+        for key, path in list(file_map.items()):
```
Comment (MetRonnie) [inline metomi/rose/upgrade.py:156]: Is this needed?

File: sphinx/api/environment-variables.rst [hunk 21]
```diff
@@ -222,17 +222,17 @@ Rose Environment Variables
 .. envvar:: ROSE_ORIG_HOST
 
    Description
-      The name of the host where the :ref:`command-rose-suite-run` command was
+      The name of the host where the :ref:`Cylc Rose` and :ref:`cylc install <Install-Workflow>`. command was
       invoked.
```
Comment (MetRonnie) [inline sphinx/api/environment-variables.rst:226]: Not sure this quite makes sense. Perhaps:
```suggestion
      The name of the host where the :ref:`cylc install <Install-Workflow>` command was
      invoked.
```

File: sphinx/tutorial/rose/furthertopics/macro-development.rst [hunk 22]
```diff
@@ -146,6 +154,8 @@ Now add the method ``_get_allowed_planets`` to the class:
                       "http://www.heavens-above.com/planetsummary.aspx"]
        p = subprocess.Popen(cmd_strings, stdout=subprocess.PIPE)
        text = p.communicate()[0]
+       if ROSEV == 2:
```
Comment (MetRonnie) [inline sphinx/tutorial/rose/furthertopics/macro-development.rst:157]:
```suggestion
       if PY_3:
```

File: sphinx/tutorial/rose/furthertopics/macro-development.rst [hunk 23]
```diff
@@ -259,41 +279,17 @@ Results
...
-re-running your macro.
-
-.. image:: img/rose-edit-macro-fail.png
-   :width: 450px
-   :align: center
-   :alt: Screenshot of setting with an error detected by a Rose macro.
+   rose macro planet.PlanetChecker
 
 Try changing the value of ``env=WORLD`` to other solar system planets
 and re-running the macro.
```
Comment (MetRonnie) [inline sphinx/tutorial/rose/furthertopics/macro-development.rst:290]: Duplicate sentence
```suggestion
```

File: sphinx/tutorial/rose/furthertopics/trigger.rst [hunk 24]
```diff
@@ -102,7 +102,7 @@ Let's add some trigger information.
 In the :rose:file:`rose-meta.conf` file, under
 ``[namelist:pizza_order=pizza_type]``, add:
 
-.. code-block:: cylc
+.. code-block:: rose
```
Comment (MetRonnie) [inline sphinx/tutorial/rose/furthertopics/trigger.rst:105]: Is there a sphinx language for rose?

File: sphinx/tutorial/rose/suites.rst [hunk 25]
```diff
@@ -402,7 +267,7 @@ See the :ref:`Cheat Sheet` for more information.
 
       .. code-block:: diff
 
-          [cylc]
+          [scheduler]
               UTC mode = True
               [[parameters]]
```
Comment (MetRonnie) [inline sphinx/tutorial/rose/suites.rst:272]: This has changed to `flow.cylc[task parameters]`

File: metomi/rose/upgrade.py (hunk 20 above)
Comment (wxtim) [reply metomi/rose/upgrade.py:156]: Can't see why it was.

File: sphinx/tutorial/rose/suites.rst (hunk 25 above)
Comment (wxtim) [reply sphinx/tutorial/rose/suites.rst:272]: Yes.

File: sphinx/tutorial/rose/furthertopics/trigger.rst (hunk 24 above)
Comment (wxtim) [reply sphinx/tutorial/rose/furthertopics/trigger.rst:105]: `rose/sphinx/ext/rose_lang.py`

**Review by MetRonnie — APPROVED** [review]

**Review by datamel — APPROVED** [review]
Thanks @wxtim.

**Pull request merged**